### ISSUE TITLE
Towards self-explainable graph neural network

### ISSUE THREAD

**Issue by 2nazero** [AIML-K/GNN_Survey#15]

**Comment by 2nazero** [comment]
https://github.com/enyandai/segnn

**Comment by 2nazero** [comment]
## Overall Summary
The paper addresses the need for self-explainable Graph Neural Network that not only predicts labels for nodes in a graph but also provides clear explanations for why certain nodes are classified as they are.

Traditional post-hoc methods attempt to explain GNN predictions after the model has made them, which may lead to explanations that don’t fully or accurately reflect the true reasoning behind predictions.

## Contributions: SE-GNN (Self-Explainable GNN)
- L-Nearest Labeled Nodes: SE-GNN identifies the closest L labeled nodes for each target node, allowing the model to base predictions on the similarity of features and structures in the local graph.
- Structural Similarity Matching: SE-GNN performs edge matching to calculate the structural similarity between nodes, providing more robust explanations based on graph topology.
- Contrastive Learning: By using contrastive learning, SE-GNN distinguishes similar nodes from dissimilar ones, enhancing the model's ability to interpret node relationships.

**Comment by 2nazero** [comment]
## Methodology: : SE-GNN Architecture and Techniques
### Node Similarity
Node similarity captures the feature similarity of nodes. SE-GNN first uses an MLP to learn each node’s initial feature embeddings $X$, followed by one GCN layer to aggregate information from neighbors.

$$H_W = \text{MLP}(X), \quad H = \sigma(\tilde{A} H_W W) + H_W$$

$\tilde{A} = D^{-1/2} (A + I) D^{-1/2}$ is the normalized adjacency matrix and $H$ is the final node embedding.
   - Node similarity between nodes $v_i$ and $v_j$ is then calculated using a similarity function, such as cosine similarity:

$$S_N(v_i, v_j) = \text{cosine}(h_i, h_j)$$

**Comment by 2nazero** [comment]
## Experiments

**Comment by 2nazero** [comment]
<img width="760" alt="스크린샷 2024-11-07 오후 3 17 27" src="https://github.com/user-attachments/assets/5ef72aa5-8acd-452a-940b-a0d7344df3c9">

```
root@e69f0e0aedb4:~/SEGNN# bash train_real.sh
Namespace(K=20, T=1.0, alpha=0.5, attr_mask=0.15, batch_size=32, beta1=0.01, beta2=0.01, cuda=False, dataset='Pubmed', debug=True, epochs=200, hidden=64, hop=2, init=False, lr=0.01, model='DeGNN', nlayer=3, no_cuda=False, seed=12, weight_decay=0.0005)
=== training model ===
Epoch 1, cls loss: 1.1331895589828491 cont_loss: 0.0806 acc_val: 0.6980 best_acc_val: 0.6980
Epoch 2, cls loss: 0.7374804615974426 cont_loss: 0.0806 acc_val: 0.7180 best_acc_val: 0.7180
Epoch 3, cls loss: 0.6413862705230713 cont_loss: 0.0802 acc_val: 0.7400 best_acc_val: 0.7400
Epoch 4, cls loss: 0.58443284034729 cont_loss: 0.0796 acc_val: 0.7380 best_acc_val: 0.7400
Epoch 5, cls loss: 0.5255063772201538 cont_loss: 0.0786 acc_val: 0.7220 best_acc_val: 0.7400
Epoch 6, cls loss: 0.45011571049690247 cont_loss: 0.0769 acc_val: 0.7300 best_acc_val: 0.7400
Epoch 7, cls loss: 0.3780854642391205 cont_loss: 0.0744 acc_val: 0.7160 best_acc_val: 0.7400
Epoch 8, cls loss: 0.2936973571777344 cont_loss: 0.0707 acc_val: 0.7040 best_acc_val: 0.7400
Epoch 9, cls loss: 0.20843859016895294 cont_loss: 0.0676 acc_val: 0.6920 best_acc_val: 0.7400
Epoch 10, cls loss: 0.13450011610984802 cont_loss: 0.0637 acc_val: 0.6920 best_acc_val: 0.7400
Epoch 11, cls loss: 0.0759669616818428 cont_loss: 0.0609 acc_val: 0.6940 best_acc_val: 0.7400
Epoch 12, cls loss: 0.03820379450917244 cont_loss: 0.0589 acc_val: 0.6960 best_acc_val: 0.7400
Epoch 13, cls loss: 0.017903033643960953 cont_loss: 0.0542 acc_val: 0.6940 best_acc_val: 0.7400
...
```

**Comment by 2nazero** [comment]
# Code Comparison 3
<img width="382" alt="스크린샷 2024-11-07 오후 3 28 49" src="https://github.com/user-attachments/assets/39dedbad-4005-4ed0-947b-734feed0a68b">

```
root@e69f0e0aedb4:~/SEGNN# bash train_BAshape.sh
/opt/conda/lib/python3.8/site-packages/torch/cuda/__init__.py:52: UserWarning: CUDA initialization: Found no NVIDIA driver on your system. Please check that you have an NVIDIA GPU and installed a driver from http://www.nvidia.com/Download/index.aspx (Triggered internally at  /pytorch/c10/cuda/CUDAFunctions.cpp:100.)
  return torch._C._cuda_getDeviceCount() > 0
Namespace(K=10, T=1, alpha=0.5, attr_mask=0.5, beta1=1, beta2=1, cuda=False, debug=True, dropout=0.5, epochs=100, hidden=128, hop=1, init=False, lr=0.01, model='DeGNN', nlayer=2, no_cuda=False, sample_size=128, seed=11, weight_decay=0.0005)
=== training model ===
Epoch 1, cls loss: 0.06207555532455444 cont_loss: 15.9162 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 2, cls loss: 0.07821442931890488 cont_loss: 14.3210 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 3, cls loss: 0.10167255252599716 cont_loss: 13.8908 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 4, cls loss: 0.12009401619434357 cont_loss: 13.4723 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 5, cls loss: 0.13597513735294342 cont_loss: 11.6072 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 6, cls loss: 0.13402707874774933 cont_loss: 11.4642 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 7, cls loss: 0.14052826166152954 cont_loss: 10.2032 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 8, cls loss: 0.1241062581539154 cont_loss: 10.7389 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 9, cls loss: 0.14182981848716736 cont_loss: 11.9345 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 10, cls loss: 0.12085629999637604 cont_loss: 11.7807 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 11, cls loss: 0.15116392076015472 cont_loss: 11.1117 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 12, cls loss: 0.1615379899740219 cont_loss: 10.2897 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 13, cls loss: 0.16297703981399536 cont_loss: 10.4220 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 14, cls loss: 0.14840374886989594 cont_loss: 10.2818 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 15, cls loss: 0.12788088619709015 cont_loss: 11.4748 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 16, cls loss: 0.13422787189483643 cont_loss: 9.9767 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 17, cls loss: 0.13471050560474396 cont_loss: 10.3638 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 18, cls loss: 0.15220044553279877 cont_loss: 10.6026 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 19, cls loss: 0.14824332296848297 cont_loss: 9.7765 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 20, cls loss: 0.15234188735485077 cont_loss: 10.2859 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 21, cls loss: 0.14792807400226593 cont_loss: 10.1910 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 22, cls loss: 0.14518018066883087 cont_loss: 10.4645 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 23, cls loss: 0.14033383131027222 cont_loss: 8.9243 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 24, cls loss: 0.1567702740430832 cont_loss: 8.9019 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 25, cls loss: 0.15553423762321472 cont_loss: 9.9690 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 26, cls loss: 0.1321064978837967 cont_loss: 8.9667 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 27, cls loss: 0.1406252533197403 cont_loss: 9.8881 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 28, cls loss: 0.15064963698387146 cont_loss: 8.9274 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 29, cls loss: 0.14569571614265442 cont_loss: 9.9802 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 30, cls loss: 0.13596659898757935 cont_loss: 10.0723 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 31, cls loss: 0.12637367844581604 cont_loss: 9.4840 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 32, cls loss: 0.13335487246513367 cont_loss: 9.4027 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 33, cls loss: 0.15904490649700165 cont_loss: 9.6138 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 34, cls loss: 0.13567113876342773 cont_loss: 9.1728 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 35, cls loss: 0.14146049320697784 cont_loss: 9.5312 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 36, cls loss: 0.14910395443439484 cont_loss: 8.8298 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 37, cls loss: 0.139970064163208 cont_loss: 8.3416 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 38, cls loss: 0.15734301507472992 cont_loss: 8.9162 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 39, cls loss: 0.12262474000453949 cont_loss: 9.2629 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 40, cls loss: 0.13208478689193726 cont_loss: 8.9379 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 41, cls loss: 0.15179654955863953 cont_loss: 9.3871 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 42, cls loss: 0.1950327455997467 cont_loss: 9.0778 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 43, cls loss: 0.14594081044197083 cont_loss: 9.6815 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 44, cls loss: 0.12957167625427246 cont_loss: 8.6946 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 45, cls loss: 0.12479524314403534 cont_loss: 9.0008 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 46, cls loss: 0.1186176985502243 cont_loss: 9.1751 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 47, cls loss: 0.1334373652935028 cont_loss: 8.6731 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 48, cls loss: 0.11730404198169708 cont_loss: 9.1476 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 49, cls loss: 0.11023597419261932 cont_loss: 9.3674 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 50, cls loss: 0.13256734609603882 cont_loss: 9.8699 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 51, cls loss: 0.10870947688817978 cont_loss: 9.7029 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 52, cls loss: 0.14733000099658966 cont_loss: 8.5369 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 53, cls loss: 0.11812926828861237 cont_loss: 9.6668 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 54, cls loss: 0.11624585837125778 cont_loss: 8.9259 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 55, cls loss: 0.1154971644282341 cont_loss: 9.1312 acc_val: 1.0000 best_acc_val: 1.0000
Epoch 56, cls loss: 0.10689926892518997 cont_loss: 8.3378 acc_val: 1.0000 best_acc_val: 1.0000
...
```

**Comment by 2nazero** [comment]
# Code Comparison 4
<img width="413" alt="스크린샷 2024-11-07 오후 3 33 42" src="https://github.com/user-attachments/assets/73ae39a8-dc7d-4ac7-acca-19a5f3dcf53f">

```
root@e69f0e0aedb4:~/SEGNN# bash test_real.sh
Testing
Accuracy: 0.8000, mAP: 0.7930
cls test results: accuracy= 0.7860
Accuracy of GCN-K: 0.7820, Accuracy of CLS: 0.7860, MAP: 0.7952
cls test results: accuracy= 0.7580
Accuracy of GIN-K: 0.7560, Accuracy of CLS: 0.7580, MAP: 0.7749
cls test results: accuracy= 0.7370
Accuracy of MLP-K: 0.7390, Accuracy of CLS: 0.7370, MAP: 0.7458
Loading citeseer dataset...
Selecting 1 largest connected components
Testing
Accuracy: 0.7494, mAP: 0.7124
cls test results: accuracy= 0.7340
Accuracy of GCN-K: 0.7210, Accuracy of CLS: 0.7340, MAP: 0.7324
cls test results: accuracy= 0.6961
Accuracy of GIN-K: 0.6659, Accuracy of CLS: 0.6961, MAP: 0.6857
cls test results: accuracy= 0.6286
Accuracy of MLP-K: 0.6315, Accuracy of CLS: 0.6286, MAP: 0.6525
Testing
...
```

**Comment by 2nazero** [comment]
# Code Comparison 5

These are the results that aren't provided in the code. (Apparently, some results aren't made out of coding so...)

<img width="376" alt="스크린샷 2024-11-07 오후 3 35 13" src="https://github.com/user-attachments/assets/7f4b9698-4121-415d-879c-42a43f3d6cba">

<img width="401" alt="스크린샷 2024-11-07 오후 3 36 27" src="https://github.com/user-attachments/assets/af44a695-c42c-47e4-a9b5-50316a6c9622">

<img width="405" alt="스크린샷 2024-11-07 오후 3 36 34" src="https://github.com/user-attachments/assets/4ba69d31-0ae6-499b-815d-27f62d1be1e9">
